File: java project/project/utils/UniqueIDGenerator.java
package project.utils;

import java.util.List;

/**
 * Generates unique IDs for complaints.
 * This class keeps ID creation simple and file-based.
 */
public class UniqueIDGenerator {

    private static final String COMPLAINT_FILE = "project/data/complaints.txt";
    private static final String PREFIX = "CMP";
    private static final int START_NUMBER = 1001;

    /**
     * Generates a new unique complaint ID by scanning the complaint file.
     * If the file is empty, the first ID will be CMP1001.
     *
     * @return unique complaint ID
     */
    public static String generateComplaintId() {
        FileManager.ensureFileExists(COMPLAINT_FILE);
        List<String> lines = FileManager.readAllLines(COMPLAINT_FILE);

        int maxNumber = START_NUMBER - 1;

        for (String line : lines) {
            if (line == null || line.trim().isEmpty()) {
                continue;
            }

            String[] parts = line.split("\\|", -1);
            if (parts.length > 0) {
                String id = parts[0].trim();
                if (id.startsWith(PREFIX)) {
                    try {
                        int number = Integer.parseInt(id.substring(PREFIX.length()));
                        if (number > maxNumber) {
                            maxNumber = number;
                        }
                    } catch (NumberFormatException ignored) {
                        // Ignore badly formatted IDs and continue searching.
                    }
                }
            }
        }

        return PREFIX + (maxNumber + 1);
    }

    /**
     * Checks whether a complaint ID already exists in the complaint file.
     *
     * @param complaintId complaint ID to check
     * @return true if the ID exists
     */
    public static boolean complaintIdExists(String complaintId) {
        FileManager.ensureFileExists(COMPLAINT_FILE);
        List<String> lines = FileManager.readAllLines(COMPLAINT_FILE);

        for (String line : lines) {
            if (line == null || line.trim().isEmpty()) {
                continue;
            }
            String[] parts = line.split("\\|", -1);
            if (parts.length > 0 && parts[0].trim().equalsIgnoreCase(complaintId)) {
                return true;
            }
        }
        return false;
    }
}

File: cpp_project/utils/UniqueIDGenerator.cpp
#include "UniqueIDGenerator.h"
#include "FileManager.h"
#include <vector>
#include <sstream>
#include <iostream>

namespace project {
namespace utils {

const std::string UniqueIDGenerator::COMPLAINT_FILE = "data/complaints.txt";
const std::string UniqueIDGenerator::PREFIX = "CMP";

std::string UniqueIDGenerator::generateComplaintId() {
    FileManager::ensureFileExists(COMPLAINT_FILE);
    std::vector<std::string> lines = FileManager::readAllLines(COMPLAINT_FILE);

    int maxNumber = START_NUMBER - 1;

    for (const std::string& line : lines) {
        if (line.empty()) continue;

        std::vector<std::string> parts;
        std::stringstream ss(line);
        std::string part;
        while (std::getline(ss, part, '|')) {
            parts.push_back(part);
        }

        if (!parts.empty()) {
            std::string id = parts[0];
            if (id.find(PREFIX) == 0) {
                try {
                    int number = std::stoi(id.substr(PREFIX.length()));
                    if (number > maxNumber) {
                        maxNumber = number;
                    }
                } catch (const std::exception& e) {
                    // Ignore badly formatted IDs and continue searching.
                }
            }
        }
    }

    return PREFIX + std::to_string(maxNumber + 1);
}

bool UniqueIDGenerator::complaintIdExists(const std::string& complaintId) {
    FileManager::ensureFileExists(COMPLAINT_FILE);
    std::vector<std::string> lines = FileManager::readAllLines(COMPLAINT_FILE);

    for (const std::string& line : lines) {
        if (line.empty()) continue;
        
        std::vector<std::string> parts;
        std::stringstream ss(line);
        std::string part;
        while (std::getline(ss, part, '|')) {
            parts.push_back(part);
        }
        
        if (!parts.empty() && parts[0] == complaintId) {
            return true;
        }
    }
    return false;
}

}
}
